lib/resty/upstream/healthcheck.lua

```diff
@@ -199,6 +208,25 @@ local function peer_ok(ctx, is_backup, id, peer)
     end
 end
 
+--- This method is used to detect if a http message body matches the expected body specified
+-- in an upstream healtcheck. If the expected_body is a contains expression (%) then it tries
+-- to read from socket till it gets the string from %%. Otherwise it does an exact read of bytes
+-- and compare the result with the expected body.
+local function is_expectedbody(socket, expected_body, http_version)
+    local readuntilnew = socket:receiveuntil(http_descriptor[http_version]["delimiter_body"])
+    local headers = readuntilnew()
+
+    if not string.find(expected_body, "%%") then
+        local body = socket:receive(string.len(expected_body))
+
+        return body == expected_body
+    end
+
+    expected_body = string.gsub(expected_body, "%%", "")
+
+    return socket:receiveuntil(expected_body)()    
```

Comment:
I don't think this is the right thing to do. Since it introduces unnecessary delay (till timeout) if the body does not contain the expected pattern, which is rather wasteful. Also, this approach will fail when the chunked encoding is used and a chunk boundary happens to split the part matching the pattern.